File: .claude/skills/mt-add-image/scripts/find-substack-post.js
#!/usr/bin/env node
// Find which Substack post embeds a given image UUID, and extract a label.
// Usage: node find-substack-post.js --uuid <uuid> [--deep]
// Output: JSON { found, source?, publication?, postTitle?, postUrl?, caption?, scanned? }
//
// Strategy: RSS feed first (fast, ~recent weeks). With --deep, fall back to a
// sitemap crawl up to ~3 months back (added in the deep-lookup phase).
// A Substack CDN URL does not encode its publication, so we search each
// publication listed in config/substack-publications.json.

const fs = require("fs");
const path = require("path");

const args = process.argv.slice(2);
function flag(name) {
  const i = args.indexOf(name);
  return i !== -1 ? (args[i + 1] && !args[i + 1].startsWith("--") ? args[i + 1] : true) : undefined;
}
const uuid = flag("--uuid");
const deep = !!flag("--deep");

if (!uuid || uuid === true) {
  console.error("Usage: node find-substack-post.js --uuid <uuid> [--deep]");
  process.exit(1);
}

const CONFIG = path.resolve(__dirname, "../config/substack-publications.json");
function loadPublications() {
  try {
    return JSON.parse(fs.readFileSync(CONFIG, "utf-8"));
  } catch {
    return ["blog.bytebytego.com"];
  }
}

async function fetchText(url) {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 10000);
  try {
    const res = await fetch(url, {
      redirect: "follow",
      signal: controller.signal,
      headers: { "User-Agent": "Mozilla/5.0" },
    });
    return res.ok ? await res.text() : "";
  } catch {
    return "";
  } finally {
    clearTimeout(timeout);
  }
}

function decodeEntities(s) {
  return s
    .replace(/&amp;/g, "&")
    .replace(/&lt;/g, "<")
    .replace(/&gt;/g, ">")
    .replace(/&quot;/g, '"')
    .replace(/&#0?39;|&apos;/g, "'")
    .replace(/&nbsp;/g, " ")
    .replace(/&#(\d+);/g, (_, n) => String.fromCharCode(+n))
    .trim();
}

function stripTags(html) {
  return decodeEntities(html.replace(/<[^>]+>/g, "")).replace(/\s+/g, " ").trim();
}

// Pull the first <title> (CDATA or plain) and <link> from an <item> chunk.
function itemTitle(item) {
  const m = item.match(/<title>(?:<!\[CDATA\[)?([\s\S]*?)(?:\]\]>)?<\/title>/);
  return m ? decodeEntities(m[1].trim()) : "";
}
function itemLink(item) {
  const m = item.match(/<link>(?:<!\[CDATA\[)?([\s\S]*?)(?:\]\]>)?<\/link>/);
  return m ? m[1].trim() : "";
}

// Extract candidate topic titles from a post's TOC bullet list. ByteByteGo does
// not attach captions to images — the topic titles live only in the "in this
// issue" bullets. We can't reliably map image→title automatically (sponsor /
// video items interleave and break positional order), so we surface these
// candidates for the user to pick from. Light filtering keeps the list short:
// dedupe, drop sub-point explanations ("Term: long sentence") and over-long
// lines. The correct title is always present; the user selects it.
function extractCandidates(html) {
  const seen = new Set();
  const out = [];
  const re = /<li[^>]*>\s*(?:<p[^>]*>)?([\s\S]*?)(?:<\/p>)?\s*<\/li>/g;
  let m;
  while ((m = re.exec(html))) {
    const text = stripTags(m[1]);
    if (!text) continue;
    if (text.length < 6 || text.length > 70) continue;     // titles are short; long lines are sub-point explanations
    const key = text.toLowerCase();
    if (seen.has(key)) continue;                           // content is duplicated in the page
    seen.add(key);
    out.push(text);
  }
  return out;
}

// If the UUID sits inside a <figure>…</figure>, return that figure's
// <figcaption> text. Cover images live in <enclosure> (no figure) → "".
function captionForUuid(item, id) {
  const at = item.indexOf(id);
  if (at === -1) return "";
  const figStart = item.lastIndexOf("<figure", at);
  if (figStart === -1) return "";
  const figEnd = item.indexOf("</figure>", at);
  if (figEnd === -1) return "";
  const figure = item.slice(figStart, figEnd);
  const cap = figure.match(/<figcaption[^>]*>([\s\S]*?)<\/figcaption>/);
  return cap ? stripTags(cap[1]) : "";
}

// Extract a post title from server-rendered post HTML (og:title preferred).
function postTitleFromHtml(html) {
  const og = html.match(/<meta[^>]+property=["']og:title["'][^>]+content=["']([^"']+)["']/i);
  if (og) return decodeEntities(og[1]);
  const h1 = html.match(/<h1[^>]*>([\s\S]*?)<\/h1>/i);
  if (h1) return stripTags(h1[1]);
  const t = html.match(/<title>([\s\S]*?)<\/title>/i);
  return t ? decodeEntities(t[1].trim()) : "";
}

// Deep fallback: crawl the sitemap back ~3 months, fetch posts most-recent-first
// (capped), and look for the UUID. Heavier than RSS — only used on RSS miss.
async function searchSitemap(publication, id, maxFetch = 40, monthsBack = 3) {
  const xml = await fetchText(`https://${publication}/sitemap.xml`);
  if (!xml) return { hit: null, scanned: 0, cutoff: null };

  const cutoff = new Date();
  cutoff.setMonth(cutoff.getMonth() - monthsBack);

  const candidates = [];
  for (const block of xml.split("<url>").slice(1)) {
    const loc = block.match(/<loc>([^<]+)<\/loc>/);
    const lastmod = block.match(/<lastmod>([^<]+)<\/lastmod>/);
    if (!loc || !lastmod) continue;
    if (!/\/p\//.test(loc[1])) continue; // posts only
    const when = new Date(lastmod[1]);
    if (isNaN(when) || when < cutoff) continue;
    candidates.push({ url: loc[1], when });
  }
  candidates.sort((a, b) => b.when - a.when);

  let scanned = 0;
  for (const c of candidates.slice(0, maxFetch)) {
    scanned++;
    const html = await fetchText(c.url);
    if (!html || !html.includes(id)) continue;
    return {
      hit: {
        found: true,
        source: "sitemap",
        publication,
        postTitle: postTitleFromHtml(html),
        postUrl: c.url,
        caption: captionForUuid(html, id),
        candidates: extractCandidates(html),
      },
      scanned,
      cutoff: cutoff.toISOString().slice(0, 10),
    };
  }
  return { hit: null, scanned, cutoff: cutoff.toISOString().slice(0, 10) };
}

async function searchRss(publication, id) {
  const xml = await fetchText(`https://${publication}/feed`);
  if (!xml) return null;
  const items = xml.split("<item>").slice(1);
  for (const item of items) {
    if (!item.includes(id)) continue;
    const postTitle = itemTitle(item);
    const caption = captionForUuid(item, id);
    return {
      found: true,
      source: "rss",
      publication,
      postTitle,
      postUrl: itemLink(item),
      caption,
      candidates: extractCandidates(item),
    };
  }
  return null;
}

async function main() {
  const publications = loadPublications();
  for (const pub of publications) {
    const hit = await searchRss(pub, uuid);
    if (hit) return console.log(JSON.stringify(hit, null, 2));
  }

  // Deep fallback: sitemap crawl up to ~3 months back.
  if (deep) {
    let totalScanned = 0;
    let lastCutoff = null;
    for (const pub of publications) {
      const { hit, scanned, cutoff } = await searchSitemap(pub, uuid);
      totalScanned += scanned;
      lastCutoff = cutoff;
      if (hit) return console.log(JSON.stringify(hit, null, 2));
    }
    return console.log(JSON.stringify({
      found: false,
      source: "sitemap",
      scanned: totalScanned,
      cutoff: lastCutoff,
    }, null, 2));
  }
  console.log(JSON.stringify({ found: false }, null, 2));
}

main();

File: scripts/newsletter/find_substack_post.go
// Find which Substack post embeds a given image UUID, and extract a label.
// Usage: go run ./scripts/newsletter find-substack-post --uuid <uuid> [--deep]
// Output on hit:  JSON { found:true, source, publication, postTitle, postUrl, caption, candidates }
// Output on miss: JSON { found:false } (RSS) or { found:false, source:"sitemap", scanned, budget, cutoff } (--deep)
//
// Strategy: RSS feed first (fast, ~recent weeks). With --deep, fall back to a
// heavier sitemap crawl up to ~3 months back — opt-in because it fetches many
// posts. A Substack CDN URL does not encode its publication, so we search each
// publication listed in config/substack-publications.json (embedded at build
// time; `go run` recompiles on change, so editing the JSON still takes effect).
package main

import (
	_ "embed"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"
)

//go:embed config/substack-publications.json
var publicationsJSON []byte

func loadPublications() []string {
	var pubs []string
	if err := json.Unmarshal(publicationsJSON, &pubs); err != nil || len(pubs) == 0 {
		return []string{"blog.bytebytego.com"}
	}
	return pubs
}

// fetchPage: body text with a browser-ish UA, or "" on any error.
func fetchPage(target string) string {
	return fetchTextOK(target, 10*time.Second, "Mozilla/5.0")
}

// Total post fetches allowed across ALL publications during a --deep crawl.
const deepFetchBudget = 40

type postHit struct {
	Found       bool     `json:"found"`
	Source      string   `json:"source"`
	Publication string   `json:"publication"`
	PostTitle   string   `json:"postTitle"`
	PostURL     string   `json:"postUrl"`
	Caption     string   `json:"caption"`
	Candidates  []string `json:"candidates"`
}

var (
	locRe     = regexp.MustCompile(`<loc>([^<]+)</loc>`)
	lastmodRe = regexp.MustCompile(`<lastmod>([^<]+)</lastmod>`)
)

func parseLastmod(s string) (time.Time, bool) {
	for _, layout := range []string{time.RFC3339, "2006-01-02"} {
		if t, err := time.Parse(layout, s); err == nil {
			return t, true
		}
	}
	return time.Time{}, false
}

// searchSitemap is the deep fallback: crawl the sitemap back ~3 months, fetch
// posts most-recent-first (up to maxFetch from the shared budget), and look
// for the UUID. Heavier than RSS — only used on RSS miss.
// cutoff is "" when the sitemap itself could not be fetched.
func searchSitemap(publication, id string, maxFetch int) (hit *postHit, scanned int, cutoff string) {
	xml := fetchPage("https://" + publication + "/sitemap.xml")
	if xml == "" {
		return nil, 0, ""
	}

	cutoffTime := time.Now().UTC().AddDate(0, -3, 0)
	cutoff = cutoffTime.Format("2006-01-02")

	type candidate struct {
		url  string
		when time.Time
	}
	var candidates []candidate
	for _, block := range strings.Split(xml, "<url>")[1:] {
		loc := locRe.FindStringSubmatch(block)
		lastmod := lastmodRe.FindStringSubmatch(block)
		if loc == nil || lastmod == nil {
			continue
		}
		if !strings.Contains(loc[1], "/p/") { // posts only
			continue
		}
		when, ok := parseLastmod(lastmod[1])
		if !ok || when.Before(cutoffTime) {
			continue
		}
		candidates = append(candidates, candidate{url: loc[1], when: when})
	}
	sort.SliceStable(candidates, func(i, j int) bool {
		return candidates[i].when.After(candidates[j].when)
	})

	if len(candidates) > maxFetch {
		candidates = candidates[:maxFetch]
	}
	for _, c := range candidates {
		scanned++
		html := fetchPage(c.url)
		if html == "" || !strings.Contains(html, id) {
			continue
		}
		return &postHit{
			Found:       true,
			Source:      "sitemap",
			Publication: publication,
			PostTitle:   postTitleFromHTML(html),
			PostURL:     c.url,
			Caption:     captionForUUID(html, id),
			Candidates:  extractCandidates(html),
		}, scanned, cutoff
	}
	return nil, scanned, cutoff
}

func searchRSS(publication, id string) *postHit {
	xml := fetchPage("https://" + publication + "/feed")
	if xml == "" {
		return nil
	}
	for _, item := range strings.Split(xml, "<item>")[1:] {
		if !strings.Contains(item, id) {
			continue
		}
		return &postHit{
			Found:       true,
			Source:      "rss",
			Publication: publication,
			PostTitle:   itemTitle(item),
			PostURL:     itemLink(item),
			Caption:     captionForUUID(item, id),
			Candidates:  extractCandidates(item),
		}
	}
	return nil
}

func runFindSubstackPost(args []string) {
	fs := flag.NewFlagSet("find-substack-post", flag.ExitOnError)
	uuid := fs.String("uuid", "", "Substack S3 image uuid to search for")
	deep := fs.Bool("deep", false, "fall back to a sitemap crawl (~3 months back)")
	_ = fs.Parse(args)

	if *uuid == "" {
		fmt.Fprintln(os.Stderr, "Usage: go run ./scripts/newsletter find-substack-post --uuid <uuid> [--deep]")
		os.Exit(1)
	}

	publications := loadPublications()
	for _, pub := range publications {
		if hit := searchRSS(pub, *uuid); hit != nil {
			printJSON(hit)
			return
		}
	}

	// Deep fallback: sitemap crawl up to ~3 months back, sharing one global
	// fetch budget across all publications so coverage can't blow up as the
	// publications list grows.
	if *deep {
		totalScanned := 0
		var lastCutoff *string
		for _, pub := range publications {
			remaining := deepFetchBudget - totalScanned
			if remaining <= 0 {
				break
			}
			hit, scanned, cutoff := searchSitemap(pub, *uuid, remaining)
			totalScanned += scanned
			if cutoff != "" {
				lastCutoff = &cutoff
			}
			if hit != nil {
				printJSON(hit)
				return
			}
		}
		printJSON(struct {
			Found   bool    `json:"found"`
			Source  string  `json:"source"`
			Scanned int     `json:"scanned"`
			Budget  int     `json:"budget"`
			Cutoff  *string `json:"cutoff"`
		}{false, "sitemap", totalScanned, deepFetchBudget, lastCutoff})
		return
	}

	printJSON(struct {
		Found bool `json:"found"`
	}{false})
}
